hub/api/dataset.py

```diff
@@ -0,0 +1,69 @@
+from hub.util.slice import merge_slices
+from typing import Union
+import numpy as np
+
+
+class Dataset:
+    def __init__(self, path: str, mode: str = "a", ds_slice: slice = slice(None)):
+        """Initialize a new or existing dataset.
+
+        Args:
+            path (str): The location of the dataset.
+                Can be a local path, or a url to a cloud storage provider.
+            mode (str, optional): Mode in which the dataset is opened.
+                Supported modes include ("r", "w", "a") plus an optional "+" suffix.
+                Defaults to "a".
+            ds_slice (slice, optional): The slice object restricting the view
+                of this dataset's tensors. Defaults to slice(None, None, None).
+                Used internally for iteration.
+        """
+        self.path = path
+        self.mode = mode
+        self.slice = ds_slice
+        self.tensors = {}  # TODO: read metadata and initialize tensors
+
+    def __len__(self):
+        """Return the greatest length of tensors"""
+        return max(map(len, self.tensors.values()), default=0)
+
+    def __getitem__(self, item: Union[slice, str]):
+        if isinstance(item, str):
+            return self.tensors[item]  # TODO: throw a pretty error
+        elif isinstance(item, slice):
+            new_slice = merge_slices(self.slice, item)
+            return Dataset(self.path, self.mode, new_slice)
+        else:
+            return None  # TODO: throw a pretty error
+
+    def __setitem__(self, item: Union[slice, str], value):
+        if isinstance(item, str):
+            if isinstance(value, np.array):
+                # TODO: write data and create tensor
+                # self.tensors[item] = Tensor(...)
+                return self.tensors[item]
+            else:
+                return None  # TODO: throw a pretty error
+        else:
+            return None  # TODO: throw a pretty error
+
+    def __iter__(self):
+        for i in range(len(self)):
+            yield self[i]
+
+    @staticmethod
+    def from_path(path: str):
+        """Create a local hub dataset from unstructured data.
```

Comment:
I think this actually loads the dataset which is present at the given path and not converts a dataset to hub format.

Reply:
We will have to write a different API that converts an unstructured to hub format.

Reply:
This was taken directly from hub 1.0.

In our current model, a dataset can only exist if its data is stored somewhere in hub format.

We can get around this by returning an object that symbolizes the compute (i.e. transforms), but then it doesn't return a dataset. Either way, we just want to be consistent

Reply:
Not sure what you mean here? 

The description says "Create a local hub dataset from unstructured data." But that is not what we do here actually. We should load the hub dataset stored at the path. So may be we can change the description?

Reply:
I did not believe that this was the desired functionality. I thought that the constructor handles locally loading a dataset, like:
```
ds = hub.Dataset("./my/local/path")
```

and from_path is reserved for converting local unstructured data (using hub.auto) into hub format.
```
ds = hub.Dataset.from_path("./path/to/some/images")
```

This is my reference: https://docs.activeloop.ai/en/latest/concepts/dataset.html#auto-usage
See also: https://activeloop.atlassian.net/wiki/spaces/HP/pages/79822931/Hub+2.0+Adopt+API+design
and https://activeloop.atlassian.net/wiki/spaces/OS/pages/128483386/API